Gobierto People: Person Blog management back-end and UI, and final touches

danguita opened the pull request:
Connects to #190.

### Review notes

These changes depend on #221.

### What does this PR do?

This PR extends the `GobiertoPeople::Person` data model by adding the Person Post resource, which allows managing Post entries in a Person context. Every Post can have a rich text formatted body field, taking advantage of the `rich_text_area` `FormBuilder` helper (see https://github.com/PopulateTools/gobierto/pull/211). It can be categorised through a list of tags or keywords as well, and they are being stored within the `PersonPost` record. Also, it is providing a simple API to query `PersonPost`s by one or multiple tags.

Finally, this one wraps up some details on the Person resource integration by making People index dynamic, and setting up the deletion strategy.

### How should this be manually tested?

Having `http://gobierto.dev/admin/people/people/<person_id>/blog` (e.g. http://gobierto.dev/admin/people/people/933465536/blog) as the main entry point for this feature, let's check that, after running migrations and re-seeding the database:

- [x] All Person's Posts are listed in the index view.
- [x] A Post can be Added or Edited in this Person's context.
- [x] The Post body field can be managed in rich text format.
- [x] It is also allowed to upload files through the body field.
- [x] The tags field can be managed as a keyword list, separated by comma.

Regarding the final touches on the People view (http://gobierto.dev/admin/people/people):

- [x] All Events, Statements and Posts counters are being calculated dynamically.

---

This is how the UI looks like:

![screen shot 2017-01-09 at 2 43 26 pm](https://cloud.githubusercontent.com/assets/126392/21768481/2f51e588-d67a-11e6-8451-6d3f27e51531.jpg)

---

![screen shot 2017-01-09 at 2 44 09 pm](https://cloud.githubusercontent.com/assets/126392/21768486/33dd6b90-d67a-11e6-88ac-6522392481be.jpg)

danguita committed:
Implement `GobiertoPeople::PersonEvent` data model

danguita committed:
Implement `PersonEvent` management back-end and UI starting point

danguita committed:
Implement `PersonEvent` Locations management back-end

danguita committed:
Implement `PersonEvent` Locations front-end with the Dynamic content UI

danguita committed:
Increase testing browser’s window size for a more realistic scenario

danguita committed:
Make integration testing auth helpers fault tolerant

danguita committed:
Implement `PersonEvent` Attendees management back-end and UI

danguita committed:
Replace PersonEvent `approved` with `published` state enum item

danguita committed:
Implement PersonEvent filtering by state

danguita committed:
Refactor DynamicContent JS component to scope event bindings

danguita committed:
Implement Geocomplete behavior in `PersonEvent`’s location fields

danguita committed:
Integrate air-datepicker as date/time picker control in PersonEvent form

- Include air-datepicker’s English locale for testing purposes.

danguita committed:
Slightly enhance Attendee row in `PersonEvent` form

danguita committed:
Merge branch 'stable' into 190-person-event-management

danguita committed:
Drop unused class since datetime-local input type is not used anymore

danguita committed:
Implement `GobiertoPeople::PersonStatement` data model

danguita committed:
Implement `PersonStatement` back-end and UI w/ Dynamic content support

danguita committed:
Implement `GobiertoPeople::PersonPost` data model

danguita committed:
Implement `PersonPost` management back-end and UI

danguita committed:
Make Person resources dependent on destroy

danguita committed:
Make Person counters by resource dynamic in People index

Commit:
Check tags presence

danguita commented on app/views/gobierto_admin/gobierto_people/people/person_posts/index.html.erb:
```diff
@@ -39,7 +39,9 @@
       <td>
-        <%= person_post.tags.to_sentence %>
+        <% if person_post.tags.present? %>
+          <%= person_post.tags.to_sentence %>
```
Good catch! What about using `person_post.tags.try(:to_sentence)` instead?

ferblape replied:
I'm not a big fan of `#try` and it's quite possible we'll add a text such as "No tags" in a future iteration.

danguita replied:
Alright, that smells like a premature optimization but I totally get your point. Let's leave the block 👍

danguita commented on app/controllers/gobierto_admin/gobierto_people/people_controller.rb:
```diff
@@ -6,7 +6,7 @@ class PeopleController < BaseController
 
       def index
-        @people = current_site.people.sorted
+        @people = current_site.people.includes(:events, :statements, :posts).sorted
```
@ferblape I'm refactoring this thing a little bit because it may be a performance killer. Let me please merge this one once is approved.

ferblape replied:
👍

danguita replied:
Hey, I ended up configuring a counter cache strategy for those associations to avoid messing with kind of raw queries. It looks a bit premature to me but I feel much better now. What do you think? Feel free to punch the merge button if that looks good to you!

danguita committed:
Configure counter cache strategy for Person-related resources

danguita committed:
Make use of counter cache fields when rendering the Person items summary

danguita committed:
Implement a Rake task to help resetting Person-related counter cache

It can be executed by running:

```
bundle exec rails gobierto_people:counter_cache:reset
```

It is also invoked in DB seeding process to grant consistency across
environments.

danguita committed:
Ignore counter cache fields in `PersonForm` operations

ferblape commented:
I think it's fine. Let's merge it.

danguita commented:
Thanks, mate 👍
